/*
 * Copyright contributors to Hyperledger Besu.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
package org.hyperledger.besu.sila.sil.messages.snap;

import org.hyperledger.besu.datatypes.Hash;
import org.hyperledger.besu.sila.sil.messages.GetBlockAccessListsMessageData;
import org.hyperledger.besu.sila.p2p.rlpx.wire.AbstractSnapMessageData;
import org.hyperledger.besu.sila.p2p.rlpx.wire.MessageData;

import java.math.BigInteger;

import org.apache.tuweni.bytes.Bytes;

public final class GetBlockAccessListsMessage extends AbstractSnapMessageData {

  public GetBlockAccessListsMessage(final Bytes data) {
    super(data);
  }

  public static GetBlockAccessListsMessage readFrom(final MessageData message) {
    if (message instanceof GetBlockAccessListsMessage getBlockAccessListsMessage) {
      return getBlockAccessListsMessage;
    }
    final int code = message.getCode();
    if (code != SnapV2.GET_BLOCK_ACCESS_LISTS) {
      throw new IllegalArgumentException(
          String.format("Message has code %d and thus is not a GetBlockAccessListsMessage.", code));
    }
    return new GetBlockAccessListsMessage(message.getData());
  }

  public static GetBlockAccessListsMessage create(final Iterable<Hash> blockHashes) {
    return new GetBlockAccessListsMessage(
        GetBlockAccessListsMessageData.encodeSnapRequest(blockHashes, SIZE_REQUEST));
  }

  @Override
  public int getCode() {
    return SnapV2.GET_BLOCK_ACCESS_LISTS;
  }

  public Iterable<Hash> blockHashes(final boolean withRequestId) {
    return GetBlockAccessListsMessageData.decodeSnapRequest(data, withRequestId);
  }

  public BigInteger responseBytes(final boolean withRequestId) {
    return GetBlockAccessListsMessageData.decodeSnapResponseBytes(data, withRequestId);
  }
}
